presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java

```diff
@@ -178,20 +182,24 @@ public void testYield()
                 Optional.empty());
 
         Operator operator = joinOperatorFactory.createOperator(driverContext);
-        if (operator.needsInput()) {
-            operator.addInput(probeInput.get(0));
-        }
+        assertTrue(operator.needsInput());
+        operator.addInput(probeInput.get(0));
         operator.finish();
 
-        // we will yield 4 times due to filterFunction
+        // we will yield 40 times due to filterFunction
         for (int i = 0; i < entries; i++) {
             driverContext.getYieldSignal().setWithDelay(5 * SECONDS.toNanos(1), driverContext.getYieldExecutor());
+            filterFunctionCalls.set(0);
             assertNull(operator.getOutput());
+            assertEquals(filterFunctionCalls.get(), 1, "Expected join to stop processing (yield) after calling filter function once");
             driverContext.getYieldSignal().reset();
         }
-        // yield is not going to prevent operator from producing a page for the 5th time
+        // delayed yield is not going to prevent operator from producing a page now (yield won't be forced because filter function won't be called anymore)
         driverContext.getYieldSignal().setWithDelay(5 * SECONDS.toNanos(1), driverContext.getYieldExecutor());
-        Page output = operator.getOutput();
+        Page output = null;
+        for (int i = 0; output == null && i < 5 /* expect output page to be produced within few calls to getOutput() */; i++) {
```

Comment:
I mentioned spilling (unspilling to be exact) in the commit message. Will move comment before the loop and mention spill there too